app.js

```diff
@@ -301,15 +302,39 @@ function ClientOptionsHandler(fileSources) {
     };
     this.setCompilers = compilers => {
         const blacklistedKeys = ['exe', 'versionFlag', 'versionRe', 'compilerType', 'demangler', 'objdumper',
-            'postProcess'];
+            'postProcess', 'demanglerClassFile', 'isSemVer'];
+        let semverGroups = [];
         const copiedCompilers = JSON.parse(JSON.stringify(compilers));
-        _.each(options.compilers, (compiler, compilersKey) => {
+        _.each(copiedCompilers, (compiler, compilersKey) => {
+            // Find if this compiler's group uses semver
+            if (compiler.isSemVer && !semverGroups.includes(compiler.group)) {
+                semverGroups.push(compiler.group);
+            }
             _.each(compiler, (_, propKey) => {
                 if (blacklistedKeys.includes(propKey)) {
                     delete copiedCompilers[compilersKey][propKey];
                 }
             });
         });
+        // Order those groups which have semver defined
+        const groupped = _.groupBy(copiedCompilers, compiler => compiler.group);
+        _.each(semverGroups, group => {
+            groupped[group].sort((lhc, rhc) => {
+                // Go to some length to have a valid semver version
+                const leftVersion = semverParser.valid(lhc.semver, true) ||
+                    semverParser.valid(lhc.semver + '.0', true) ||
+                    "9999999.99999.999";
+                const rightVersion = semverParser.valid(rhc.semver, true) ||
+                    semverParser.valid(rhc.semver + '.0', true) ||
+                    "9999999.99999.999";
+                return semverParser.compare(leftVersion, rightVersion);
+            });
+            for (let order = 0;order < groupped[group].length;++order) {
```

Comment:
I'm racking my brain and I can't find it, but I think there must be a simpler way to do this. The tricky bit is having the ordinal value `order` to do the maths with `length - order`. I don't think I quite follow what's going on here, either. I see the shape of it but I'm being dumb I think.

So, we're in a compiler group. And we're counting from 0 to the number of compilers in that group. Got that.

Then we find that `xthCompiler` in the big bag of copiedCompilers, by looking it up by its unique id. Got it. I don't think this can fail, right? That is, `xthCompiler` should never be falsey; else something really funky has gone on and we have invented a compiler in our grouped list that wasn't in the original. I'd probably drop the `if()` check as it confuses things (IMO).

So then we assign that compiler (the one in the `copiedCompilers`, which should be the exact same object as the `groupped[group][order]` I think?) - we assign it an `$order` of `length-order` -- presumably the `length-order` is to reverse the sort order? If we just want to reverse the order, does `-order` (a negative order) just work? Seems simpler.

OK. Sorry for all the thinking out loud, but I thought it might be handy to see how I'm thinking about this. We should pair program sometime :)

With all that thought about I _think_ we can do...

```javascript
_.each(grouped, group => {
  group[$order] = - _.findIndex(group, grouped);
});
```

I _think_, right? Each group has an $order which picks (minus) its position in its own group?

Now I'm less sure, but I'll post this anyway as I have to go make dinner and hopefully you can decipher what I meant :)

Maybe some tests would help flesh this out and lock in good behaviour?

Reply:
>nitpick: spaces after the semicolon. Ctrl-Alt-L should fix all that :)

Woops. You're right. Maybe the linter should catch that too?

> I'm racking my brain and I can't find it, but I think there must be a simpler way to do this. The tricky bit is having the ordinal value order to do the maths with length - order. I don't think I quite follow what's going on here, either. I see the shape of it but I'm being dumb I think.

Sorry for writing something so ugly that it's hard to read! 😨 

> OK. Sorry for all the thinking out loud, but I thought it might be handy to see how I'm thinking about this. We should pair program sometime :)

Sure!

> _.each(grouped, group => {
          group[$order] = - _.findIndex(group, grouped);
});

I ended up doing something a bit less terrible, where the semverGroups variable now holds the groups, with help from the `sortedIndex` function. Hopefully this way the code becomes more readable.

> I think, right? Each group has an $order which picks (minus) its position in its own group?

Note that group was an array of `compilerInfos` in this case, so we still had to loop thru each one of the items

> Maybe some tests would help flesh this out and lock in good behaviour?

Yes! They only reason they are not yet there is because creating tests at 4am is a bit more soulcrushing that what I can handle at this hours! :) I won't accept a merge until I've got time to add at least some basic tests